controllers/Handle.php

```diff
@@ -81,25 +109,39 @@ public function callback(string $provider): RedirectResponse
         }
 
         if (!$user) {
-            // $password = Str::random(400);
-            // $user = $this->authManager->register([
-            //     'email' => $ssoUser->getEmail(),
-            //     'password' => $password,
-            //     'password_confirmation' => $password,
-            //     'name' => $ssoUser->getName(),
-            // ]);
-            // $user->setSsoConfig('allow_password_auth', false);
-            // @TODO: Event here for registering user if desired, default fallback abort behaviour
-            abort(403, 'User not found');
+            Event::listen('winter.sso.register', function ($controller, $provider, $ssoUser) {
+                return;
+                $password = Str::random(400);
+                $user = $controller->getAuthManager()->register([
+                    'email' => $ssoUser->getEmail(),
+                    'password' => $password,
+                    'password_confirmation' => $password,
+                    'name' => $ssoUser->getName(),
+                ]);
+                $user->setSsoValues($provider, ['allow_password_auth' => false]);
+                return $user;
+            });
+            if (Config::get('winter.sso::allow_registration')) {
+                $user = Event::fire('winter.sso.register', [$this, $provider, $ssoUser]);
+            }
+            if (!$user) {
+                Flash::error(trans('winter.sso::lang.messages.user_not_found', ['user' => $ssoUser->getEmail()]));
+                return $this->redirectToSignInPage();
+            }
```

Comment:
Why is the user registration happening in an event listener instead of outside? If we want to support using the events to swap out the actual registration code itself then maybe we could use two separate events instead so that the register event is always the same?

Side note; I'd like any event names to fit in well with https://github.com/wintercms/winter/issues/296

Reply:
> Why is the user registration happening in an event listener instead of outside? If we want to support using the events to swap out the actual registration code itself then maybe we could use two separate events instead so that the register event is always the same?

I used the same technique we used in the database model events where we use an event that can be overriden (I forgot to use the halt:true when I fire the event though).

> Side note; I'd like any event names to fit in well with [wintercms/winter#296](https://github.com/wintercms/winter/issues/296)

Will take a look. I just used those until I could figure out the correct event names.

Reply:
@LukeTowers Still trying to wrap my head around the implications for https://github.com/wintercms/winter/issues/296 ... What does it mean for the event I named `winter.sso.signin` in this PR ? This happens when the user signedin with the SSO provider (and we get its information in the new ssoUser object)... how do you want to convert that to a model event method ?